File: Data Structures & Algorithms/buy-and-sell-crypto/submission-2.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {

        let mut min = prices[0];
        let mut max_profit =0 ;
        for &element in prices.iter(){

            min = min.min(element);

            let profit = element - min;

            max_profit = max_profit.max(profit);

        }
        max_profit
    }
}

File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-2.cpp
class Solution {
public:
    int recurse(vector<int> &prices, int i, bool holding, vector<vector<int>> &dp){
        if(i>=prices.size()){
            return 0;
        }
        if(dp[i][holding]!= -1){
            return dp[i][holding];
        }
        int hold = recurse(prices, i+1, holding,dp);
        if(!holding){
            int buy = -prices[i]+ recurse(prices, i+1, true,dp);
            return dp[i][holding]=max(hold, buy);
        }else{
            int sell = prices[i]+ recurse(prices,i+2, false,dp);
            return dp[i][holding]=max(hold, sell );
        }
    }
    int maxProfit(vector<int>& prices) {
        vector<vector<int>> dp(prices.size(), vector<int>(2,-1));
        return recurse(prices,0, false, dp);
    }
};
